algebra2/gaussian_elimination/gaussian_elimination.hpp

```diff
@@ -0,0 +1,500 @@
+    template<typename T, typename LP>
+    auto show_steps(ranges::matrix_view<T, LP> m,
+                    axis a,
+                    operation allowed_operations = operation::swap | operation::add | operation::multiply,
+                    reducted_form reducted = reducted_form::echelon) -> void{
+        std::vector<std::string> steps = gaussian_elimiantion_alg(m, axis::rows, reducted).second.first;
+        std::print("{}\n", steps);
+    }
+
+
+    /*
+        description:
+            calculates the determinant of matrix m
+    */
+    template <typename T, typename LP>
+    auto determinant(ranges::matrix_view<T, LP> m) -> std::expected<fraction<int>, error>{
+        int rows = m.extent(0),
+            cols = m.extent(1);
+
+        if(rows == cols){
+            fraction<int> det = gaussian_elimiantion_alg(m, axis::rows, reducted_form::diagonal).second.second;
+            return det;
+        }
+        else{
+            return std::unexpected(error::not_square);
+        }
+    }
+
+    /*
+        description:
+            returns inverse of matrix m
+    */ 
+    template <typename T, typename LP>
+    auto inverse(ranges::matrix_view<T, LP> m) -> std::expected<std::pair<std::vector<fraction<int>>, ranges::matrix_view<fraction<int>, LP>>, error>{
+        int rows = m.extent(0),
+            cols = m.extent(1);
+        auto [identity_vector, identity_matrix] = identity_generator(m);
+        auto [combined_vector, combined_matrix] = combine_matrices(identity_matrix, m);
+
+        if(determinant(m).value().numerator != 0){
+            auto [reduced_vector, reduced_matrix] = gaussian_elimiantion_alg(combined_matrix, axis::rows, reducted_form::diagonal).first;
+            auto [inverse_vector, inverse_matrix] = split_matrix(reduced_matrix);
+            return std::pair{std::move(inverse_vector), inverse_matrix};
+        }else{
+            return std::unexpected(error::not_invertible);
+        }
+    }
+
+}  // namespace algorithms::gaussian_elimination
+
+/*
+    description:
+        examples of using different gaussian elimination operations on matrices
+*/
+auto examples(){
+    std::vector v1{1, 1, -9, 6, 7, 10, 0, 0, -6, 4, 2, 3, 0, 0, -3, 2};
+    std::vector v2{1, -3, -3, -1, 1, 2, 2, -3, -14};
+    std::vector<algorithms::gaussian_elimination::fraction<int>> v3{{1, 2}, {1, 2}, {1, 2}, {1, 2}, {2, 4}, {1, 3}, {5, 3}, {1, 2}, {9, 4}};
+    std::vector<algorithms::gaussian_elimination::fraction<int>> v4{{1, 3}, {1, 5}, {9, 2}, {5, 6}, {6, 7}, {11, 14}, {15, 10}, {0, 1}, {9, 1}};
+
+    ranges::matrix_view m1(v1, 4, 4, layout::row);
+    ranges::matrix_view m2(v2, 3, 3, layout::row);
+    ranges::matrix_view m3(v3, 3, 3, layout::row);
+    ranges::matrix_view m4(v4, 3, 3, layout::row);
+
+    std::println("//////Gaussian Elimination Examples//////\n");
+    auto [m1_vector, m1_reduced] = algorithms::gaussian_elimination::run(m1, algorithms::gaussian_elimination::axis::rows);
+    std::println("{}\n", m1_reduced);
```

Comment:
you should add some tests (not just examples): use sth like assert(std::ranges::equal(&lt;calculated by you matrix&gt;, &lt;expected matrix&gt;))